source/common/event/guarddog_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "envoy/event/guarddog.h"
+#include "envoy/event/watchdog.h"
+#include "envoy/server/configuration.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/logger.h"
+#include "common/common/thread.h"
+#include "common/event/libevent.h"
+
+namespace Event {
+
+/** This feature performs deadlock detection stats collection&enforcement.
+ *
+ * It launches a thread that scans at an interval the minimum of the configured
+ * intervals.  If it finds starved threads or suspected deadlocks it will take
+ * the appropriate action depending on the config parameters described below.
+ *
+ * Thread lifetime is tied to GuardDog object lifetime (RAII style).
+ */
+class GuardDogImpl : public GuardDog {
+public:
+  /**
+   * @param stats_store Statistics store to write watchdog_miss and
+   * watchdog_mega_miss events into.
+   * @param config Configuration object.
+   *
+   * Configuration:
+   *   1) Miss timeout: Record a "miss" counter increment after this long has
+   *   passed.  Old default: 100ms.
+   *   2) Megamiss timeout: Record a "megamiss" counter increment after this
+   *   long has passed.  Old setting: 1 second.
+   *   3) Kill timeout: If any thread is stuck for longer than this kill the
+   *   process.  0 to disable kill.
+   *   4) Multi kill timeout: If two or more threads are stuck for longer than
+   *   this kill the process. 0 to disable.
+   *
+   */
+  GuardDogImpl(Stats::Store& stats_store, const Server::Configuration::Main& config,
+               SystemTimeSource& tsource);
+  ~GuardDogImpl();
+
+  WatchDogSharedPtr getWatchDog(int32_t thread_id) override;
+  void stopWatching(WatchDogSharedPtr wd) override;
+
+  /**
+   * Exposed for testing purposes only (but harmless to call):
+   */
+  int loopInterval() const { return loop_interval_.count(); }
+  void force_check() {
```

Comment:
Nit: `forceCheck`. I think you could also call these somethings like `loopIntervalCountforTest()`, `forceCheckForTest()` if they really are only ever to be used in testing.